internal/worker/annotator.go

```diff
@@ -44,14 +47,25 @@ func NewAnnotator(poolSize int, orasClient api.ORASClientInterface, loginURL str
 	}, nil
 }
 
-// AnnotateManifests annotates a list of digests (tags and manifests) concurrently and returns a count of annotated tags & manifests and the first error occurred.
-func (a *Annotator) Annotate(ctx context.Context, digests *[]string) (int, error) {
-	jobs := make([]job, len(*digests))
-	for i, digest := range *digests {
-		jobs[i] = newAnnotateJob(a.loginURL, a.repoName, a.artifactType, a.annotations, a.orasClient, digest)
-	}
+// AnnotateManifests annotates a list of manifests concurrently and returns a count of annotated images and the first error occurred.
+func (a *Annotator) Annotate(ctx context.Context, manifests []string) (int, error) {
+	var annotatedImages atomic.Int64
+	group := a.pool.NewGroup()
 
-	return a.process(ctx, &jobs)
+	for _, digest := range manifests {
+		group.SubmitErr(func() error {
+			ref := fmt.Sprintf("%s/%s@%s", a.loginURL, a.repoName, digest)
+			if err := a.orasClient.Annotate(ctx, ref, a.artifactType, a.annotations); err != nil {
+				fmt.Printf("Failed to annotate %s/%s@%s, error: %v\n", a.loginURL, a.repoName, digest, err)
+				return err // TODO: #469 Do we want to fail the whole job if one fails? This is the current behaviour.
```

Comment:
no I don't think we want to fail the whole job, I think the current behavior is not good - we should collect errors and continue, or handle the error (retry?) - right now the current behavior is very fragile

Reply:
I agree, but annotate is a contribution from a separate team and I have a couple of things to bring up with them so its left as a todo for now.